docs/design/vsock-design.md

```diff
@@ -0,0 +1,208 @@
+this mechanism:
+
+- It should maintain the Firecracker security barrier, isolating the guest from
+  direct access to any host resources; and
+- It should support multiple communication channels; and
+- It should be more lightweight than traditional network sockets, in the sense
+  of requiring fewer host resources. See
+  [this considered approach](#ethernet-host-to-ethernet-guest).
+
+## Context
+
+An established approach to guest ↔ host communication is provided by the VirtIO
+vsock device. In its current Linux implementation, this approach requires three
+actors: the virtio-vsock driver (provided by the guest OS), the virtio-vosck
+device control plane (provided by the VMM), and the vhost-vsock device data
+plane (provided by the host Linux kernel).
+
+Using the vhost-vsock data plane, however, means that guest code gets to
+control various inputs into the host kernel, thus obtaining a direct path to
+exploiting any potential vulnerability affecting the vhost kernel code. This
+wouldn't be compliant with the Firecracker security model - a new attack
+surface would be added, providing the guest with a potential path to break
+isolation. 
+
+At the same time, VirtIO vsock is the most prevalent communication mechanism
+employed by today's container orchestrators - an ecosystem with which
+Firecracker aims to integrate, as a secure runtime, backed by hardware
+virtualization.
+
+## Proposed Approach
+
+Our proposed solution is to provide full virtio-vsock support to software
+running inside the guest VM, while bypassing vhost kernel code on the host. To
+that end, Firecracker will implement the virtio-vsock device model, and mediate
+communication between AF_UNIX sockets (on the host end) and AF_VSOCK sockets
+(on the guest end). Both guest and host software will be able to use the
+familiar socket interface.
+
+In order to provide channel multiplexing, AF_VSOCK ports will be translated
+into multiple AF_UNIX sockets (one unix socket per vsock port). Each vsock
+device will have a unix socket path associated. E.g. `/path/to/vsock_{CID}`.
+There are two scenarios to be considered, depending on where the connection is
+initiated.
+
+### Guest-Initiated Connections
+
+When the virtio-vsock device model in Firecracker detects a connection
+request coming from the guest (a VIRTIO_VSOCK_OP_REQUEST packet), it will try
+to forward the connection to an AF_UNIX socket listening on the host, at
+`/path/to/vsock_{CID}:{PORT}`, where `{PORT}` is the destination port, as
+specified in the connection request packet. If no such socket exists, or
+no one is listening on it, a connection cannot be established, and
+a VIRTIO_VSOCK_OP_RST packet will be sent back to the guest.
+
+From the user perspective, these would be the steps taken to establish a
+communication channel:
+
+1. Host: At VM configuration time, add a virtio-vsock device with `CID` and
+   `PATH`;
+2. Host: create and listen on an AF_UNIX socket at `{PATH}_{CID}:{PORT}`;
```

Comment:
Is it required for the AF_UNIX socket to be created before the VM starts/is configured, or is the requirement that the AF_UNIX socket is created before the guest issues the `connect()` call?

Reply:
the `AF_UNIX` socket only needs to be present when the guest calls `connect()`